File: week14/29/minesweeper.py
"""
<Minesweeper>
"""

def update_board(board, click):
    rows, cols = len(board), len(board[0])
    directions = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
    
    def count_adjacent_mines(r, c):
        count = 0
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and board[nr][nc] == 'M':
                count += 1
        return count

    def dfs(r, c):
        if not (0 <= r < rows and 0 <= c < cols) or board[r][c] != 'E':
            return
        adjacent_mines = count_adjacent_mines(r, c)
        if adjacent_mines > 0:
            board[r][c] = str(adjacent_mines)
        else:
            board[r][c] = 'B'
            for dr, dc in directions:
                dfs(r + dr, c + dc)
    
    clickr, clickc = click
    if board[clickr][clickc] == 'M':
        board[clickr][clickc] = 'X'  # Game over
    else:
        dfs(clickr, clickc)
    
    return board

# Example usage:
board = [
    ["E", "E", "E", "E", "E"],
    ["E", "E", "M", "E", "E"],
    ["E", "E", "E", "E", "E"],
    ["E", "E", "E", "E", "E"]
]
click = [3, 0]

updated_board = update_board(board, click)
for row in updated_board:
    print(row)

File: week14/6/minesweeper.cpp
#include <vector>

using namespace std;

class Solution {
 public:
  bool inboard(const vector<vector<char>>& board, int x, int y) {
    return (x >= 0 && x < board.size() && y >= 0 && y < board[0].size());
  }

  void dfs(vector<vector<char>>& board, int x, int y) {
    if (!inboard(board, x, y)) return;
    if (board[x][y] == 'B') return;

    int count = 0;
    if (inboard(board, x - 1, y - 1) && board[x - 1][y - 1] == 'M') count++;
    if (inboard(board, x - 1, y) && board[x - 1][y] == 'M') count++;
    if (inboard(board, x - 1, y + 1) && board[x - 1][y + 1] == 'M') count++;
    if (inboard(board, x, y - 1) && board[x][y - 1] == 'M') count++;
    if (inboard(board, x, y + 1) && board[x][y + 1] == 'M') count++;
    if (inboard(board, x + 1, y - 1) && board[x + 1][y - 1] == 'M') count++;
    if (inboard(board, x + 1, y) && board[x + 1][y] == 'M') count++;
    if (inboard(board, x + 1, y + 1) && board[x + 1][y + 1] == 'M') count++;

    if (count > 0)
      board[x][y] = '0' + count;
    else {
      board[x][y] = 'B';
      dfs(board, x - 1, y - 1);
      dfs(board, x - 1, y);
      dfs(board, x - 1, y + 1);
      dfs(board, x, y - 1);
      dfs(board, x, y + 1);
      dfs(board, x + 1, y - 1);
      dfs(board, x + 1, y);
      dfs(board, x + 1, y + 1);
    }
  }

  vector<vector<char>> updateBoard(vector<vector<char>>& board,
                                   vector<int>& click) {
    int r = click[0];
    int c = click[1];
    if (board[r][c] == 'M') {
      board[r][c] = 'X';
      return board;
    }

    dfs(board, r, c);
    return board;
  }
};